module/zfs/zfs_vnops.c

```diff
@@ -2950,25 +2950,27 @@ zfs_setattr(struct inode *ip, vattr_t *vap, int flags, cred_t *cr)
 	if (mask & (ATTR_UID|ATTR_GID)) {
 
 		if (mask & ATTR_UID) {
-			SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_UID(zsb), NULL,
-			    &new_uid, sizeof (new_uid));
 			ZTOI(zp)->i_uid = SUID_TO_KUID(new_uid);
+			new_kuid = zfs_uid_read(ZTOI(zp));
+			SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_UID(zsb), NULL,
+			    &new_kuid, sizeof (new_kuid));
```

Comment:
I've renamed the variables altogether, what we get from the vapp struct is now the new_kuid/kgid and what we send to the SA is the new_uid, hopefully making it clear the type of values the code is working with.